src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1121,10 +1144,66 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
             falseTestPermitsTrueOther = true;
             trueTestImpliesTrueOther = false;
             falseTestImpliesTrueOther = false;
+            lengthTest = Tests.True.Instance;
 
             // if the tests are for unrelated things, there is no implication from one to the other
-            if (!test.Input.Equals(other.Input))
-                return;
+            BoundDagTemp input1 = test.Input;
```

Comment:
nit: `// if the tests are for unrelated things, there is no implication from one to the other` this comment (line 1149) is out-of-date.

Reply:
The new block below deserves some more comments too.

My understanding for the `case (BoundDagIndexerEvaluation s1, BoundDagIndexerEvaluation s2):` section:
Even though the two tests appear unrelated (different inputs) it is possible that they are in fact related under certain conditions. 
For instance, the inputs  `[0]` and `[^1]` point to the same element when length is 1.
We'll determine that condition (`lengthTest`) and proceed with analysis of how the two tests relate under that assumption.
We also normalize indexer inputs to account for indexer accesses inside slices: `[0]` in nested list-pattern `[ 0, ..[$$], 2 ]` refers to `[1]` in the top-level list-pattern.

I haven't wrapped my head around the second section (`case (BoundDagEvaluation s1, BoundDagEvaluation s2) when s1.IsEquivalentTo(s2):`) yet ;-)

Reply:
The meaning of `unrelated` isn't change, it's just extended beyond the "same input".